### PR TITLE
enable passing datetime tests on PyPy

### PR DESCRIPTION
**Pull request by mattip**
These tests are enabled on master, and passed for me locally. 

I am trying to track down pandas-dev/pandas#50817 where using a PyDateTimeDelta crashes, it [seems](https://github.com/pandas-dev/pandas/issues/50817#issuecomment-1400786923)
> `(struct __pyx_vtabstruct_6pandas_5_libs_6tslibs_10timedeltas__Timedelta *)__pyx_v_self->__pyx_vtab` is `NULL`, which results in a segfault, any hints of how to write a test for this are welcome.

### CODE REVIEW

**Comment by da-woods** [comment]
> any hints of how to write a test for this are welcome.

So looking at it you have three layers of inheritance:

1. A Python timedelta object.
2. A Cython cdef class that defines at least one cdef function (so that it has a vtab attribute).
3. A Cython-defined def class, that initializes itself via `__new__`.

The suspicion has to be that the (implicitly defined) `__cinit__`/`tp_new` for the `cdef class` isn't getting called and thus vtab isn't getting initialized. I'd start by trying to see why not, and maybe whether defining `__cinit__` explicitly changes things or at least lets you put in some debugging prints

**Comment by h-vetinari** [comment]
If it's just a matter of the cython version, I don't mind building arrow-for-PyPy with the Cython 3.0 betas. In terms of impact, I think chasing errors on 0.29 is not a good use of people's scarce time, given how close we seem to be getting to releasing 3.0 (this year 🤞)

**Comment by mattip** [comment]
TL;DR: the problem is due to PyPy, not Cython.

The problem is:
- pandas overrides the `timedelta.days` getter with a cython-based one that assumes the object is fully constructed (i.e. it has set the `__pyx_vtab`, which is done one line after calling `tp_new()` 
- pypy assumed it could safely call that getter descriptor when converting a pure-python `datetime.timedelta` into a C `PyDateTime_Delta` object with a `days` field in the C struct
- When calling `tp_new()` from C, it is not safe to use the descriptor, since `__pyx_vtab` is not set.

The solution is for PyPy to not assume it can call the `days` getter from C, rather it should use the private `_days` attribute which is well defined once the python object is created. Fixed in [267c2f5eca33](https://foss.heptapod.net/pypy/pypy/-/commit/0864e9101f241150101b7ffcad36fb7fe23d484a)

I will close this PR, the tests pass but not worth the churn. Thanks for looking.

**Comment by da-woods** [comment]
> I will close this PR, the tests pass but not worth the churn. Thanks for looking.

I'm going to claim that it is worth enabling the tests, given that the work has been done to identify that they work

**Comment by h-vetinari** [comment]
> to identify that they work

With an as-yet-unreleased version of pypy, AFAICT...

**Comment by da-woods** [comment]
I think that the unreleased version PyPy is needed to use datetimes how pandas uses them, but not needed for these specific tests. But if I'm wrong we can definitely revert it!

**Comment by mattip** [comment]
Sorry for the noise. Indeed, the tests pass locally on the released version of PyPy without the fix. The fix is needed for pandas (and so far only for pandas) since pandas has a class that overrides the python-level `days` getter in `datetime.timedelta`.
